qualification/Jenkinsfile

```diff
@@ -0,0 +1,111 @@
+
+/* This file tells Jenkins how to execute the qualification tests for this repo.
+ * Everything runs in docker containers, so in theory any Jenkins server should
+ * be able to parse this file, although the Jenkins server needs access to a GPU
+ * with tensor cores, and the pytest-jenkins.ini file needs to be configured for the
+ * hardware it is running on. The Docker Engine being used needs to have been
+ * configured to use the Nvidia Container Runtime and the node which the container
+ * runs on needs a Nvidia GPU with tensor cores along with the Nvidia Driver installed.
+ *
+ * Additionally the Jenkins server also needs access to a Mellanox NIC that
+ * supports ibverbs. The ibverbs drivers need to be passed into the Jenkins
+ * container using specific flags.
+ *
+ * For more information on the qualification framework, see katgpucbf/doc/qualification.rst
+ */
+
+pipeline {
+  agent {
+    dockerfile {
+
+      /* Use the Jenkins-specific stage of the Dockerfile as the image for
+       * testing. This provides the appropriate dependencies.
+       */
+      additionalBuildArgs '--target=jenkins'
+
+      /* The following argument needs to be specified in order for the container
+       * to launch correctly.
+       *
+       * --gpus all: Assigns all available GPUs to the container.
+       *
+       * -v /var/run/docker.sock:/var/run/docker.sock:
+       * This allows the Unix socket that the Docker daemon listens on by default, to be used
+       * by the daemon from within the container.
+       *
+       * --network=host: The Docker container requires access to the high-speed
+       * correlator network. This command passes all the host network interfaces to the
+       * container to be used as required.
+       *
+       * -e NVIDIA_MOFED=enabled: This will pass the drivers required for ibverbs
+       * to the container.
+       *
+       * --runtime=nvidia': This will pass the NVIDIA Container Runtime information
+       * to the container and allow for a GPU accelerated Docker container to be run.
+       *
+       */
+      args '--gpus=all -v /var/run/docker.sock:/var/run/docker.sock --network=host --ulimit=memlock=-1 -e NVIDIA_MOFED=enabled --runtime=nvidia'
+    }
+  }
+
+  options {
+    timeout(time: 1, unit: 'HOURS')
+  }
+
+  stages {
+
+    /* This stage executes the qualification tests on the latest docker image
+     * of katgpucbf that has been previously unit-tested, and generates a report of the
+     * qualification test results.
+     */
+    stage('Run qualification tests') {
+       steps {
+         script {
+           sh 'pip install -r qualification/requirements.txt'
+           sh 'pip install --no-deps ".[qualification]" && pip check'
+           sh "spead2_net_raw pytest -v -c qualification/pytest-jenkins.ini qualification --image-override katgpucbf:harbor.sdp.kat.ac.za/cbf/katgpucbf:latest"
+           sh "qualification/report/generate_pdf.py report.json report.pdf"
```

Comment:
Any particular reason for mixing single and double quotes?

Reply:
Before when I had a variable in this command the single quotes did not work. But now that the variable is removed I can switch back to single quotes.